core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -0,0 +1,787 @@
+  private AtomicUpdateGroup<T> findAtomicUpdateGroupWith(PartitionChunk<T> chunk, State state)
+  {
+    final Short2ObjectSortedMap<AtomicUpdateGroup<T>> versionToGroup = getStateMap(state).get(
+        RootPartitionRange.of(chunk)
+    );
+    if (versionToGroup != null) {
+      final AtomicUpdateGroup<T> atomicUpdateGroup = versionToGroup.get(chunk.getObject().getMinorVersion());
+      if (atomicUpdateGroup != null) {
+        return atomicUpdateGroup;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns null if atomicUpdateGroup is not found for the state.
+   * Can return an empty atomicUpdateGroup.
+   */
+  @Nullable
+  private AtomicUpdateGroup<T> tryRemoveChunkFromGroupWithState(PartitionChunk<T> chunk, State state)
+  {
+    final RootPartitionRange rangeKey = RootPartitionRange.of(chunk);
+    final Short2ObjectSortedMap<AtomicUpdateGroup<T>> versionToGroup = getStateMap(state).get(rangeKey);
+    if (versionToGroup != null) {
+      final AtomicUpdateGroup<T> atomicUpdateGroup = versionToGroup.get(chunk.getObject().getMinorVersion());
+      if (atomicUpdateGroup != null) {
+        atomicUpdateGroup.remove(chunk);
+        if (atomicUpdateGroup.isEmpty()) {
+          versionToGroup.remove(chunk.getObject().getMinorVersion());
+          if (versionToGroup.isEmpty()) {
+            getStateMap(state).remove(rangeKey);
+          }
+        }
+
+        determineVisibleGroupAfterRemove(
+            atomicUpdateGroup,
+            RootPartitionRange.of(chunk),
+            chunk.getObject().getMinorVersion(),
+            state
+        );
+        return atomicUpdateGroup;
+      }
+    }
+    return null;
+  }
+
+  private List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> findOvershadowedBy(
+      AtomicUpdateGroup<T> aug,
+      State fromState
+  )
+  {
+    final RootPartitionRange rangeKeyOfGivenAug = RootPartitionRange.of(aug);
+    return findOvershadowedBy(rangeKeyOfGivenAug, aug.getMinorVersion(), fromState);
+  }
+
+  private List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> findOvershadowedBy(
+      RootPartitionRange rangeOfAug,
+      short minorVersion,
+      State fromState
+  )
+  {
+    Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> current = getStateMap(fromState)
+        .floorEntry(rangeOfAug);
+
+    if (current == null) {
+      return Collections.emptyList();
+    }
+
+    // Find the first key for searching for overshadowed atomicUpdateGroup
+    while (true) {
+      final Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> lowerEntry = getStateMap(fromState)
+          .lowerEntry(current.getKey());
+      if (lowerEntry != null && lowerEntry.getKey().startPartitionId == rangeOfAug.startPartitionId) {
+        current = lowerEntry;
+      } else {
+        break;
+      }
+    }
+
+    final List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> found = new ArrayList<>();
+    while (current != null && rangeOfAug.contains(current.getKey())) {
```

Comment:
this method could maybe use a few more comments to make understanding what is going on a bit easier

Reply:
👍 added comments and javadoc.